File: java/complete/myretro/src/main/java/com/apress/myretro/retro/RetroBoardService.java
package com.apress.myretro.retro;

import com.apress.myretro.client.UserClient;
import com.apress.myretro.persistence.Repository;
import com.apress.myretro.user.User;
import lombok.AllArgsConstructor;
import org.springframework.stereotype.Service;

import java.util.ArrayList;
import java.util.Optional;
import java.util.UUID;

@AllArgsConstructor
@Service
public class RetroBoardService {
    private Repository<RetroBoard, UUID> repository;
    private UserClient userClient;

    public Iterable<RetroBoard> getAllAvailableBoars(){
        return this.repository.findAll();
    }

    RetroBoard saveBoard(RetroBoard retroBoard){
        return this.repository.save(retroBoard);
    }

    Optional<RetroBoard> findBoardById(String boardId){
        return this.repository.findById(UUID.fromString(boardId));
    }

    void deleteBoardById(String boardId){
        this.repository.delete(UUID.fromString(boardId));
    }

    Iterable<Card> getCardsFromBoardId(String boardId){
        RetroBoard board = findBoardById(boardId).get();
        if (null == board) return null;
        return board.getCards().values();
    }

    void removeCardFromBoard(String cardId, String boardId){
        RetroBoard board = findBoardById(boardId).get();
        if (null == board) return;
        board.getCards().remove(UUID.fromString(cardId));
    }

    Card saveCardTo(String boardId, Card card){
        RetroBoard board = findBoardById(boardId).get();
        if (null == board) return null;


        if (null == card.getId()){
            card.setId(UUID.randomUUID());
            board.getCards().put(card.getId(),card);
            return card;
        }

        Card updatedCard = board.getCards().get(card.getId());
        updatedCard.setComment(card.getComment());
        return updatedCard;
    }

    User assignUserToBoard(String boardId, String email){
        RetroBoard board = findBoardById(boardId).get();
        if (null == board) return null;
        User user = this.userClient.getUserByEmail(email);

        if (null != user) board.getAssignedUsers().put(user.getEmail(),user);

        return user;
    }

    // New
    Iterable<Card> getHappyCardsFromRetroBoardId(String boardId) {
        return getCardsFromBoardIdAndCardType(boardId,CardType.HAPPY);
    }

    Iterable<Card> getMehCardsFromRetroBoardId(String boardId) {
        return getCardsFromBoardIdAndCardType(boardId,CardType.MEH);
    }

    Iterable<Card> getSadCardsFromRetroBoardId(String boardId) {
        return getCardsFromBoardIdAndCardType(boardId,CardType.SAD);
    }

    private Iterable<Card> getCardsFromBoardIdAndCardType(String boardId, CardType cardType){
        RetroBoard board = findBoardById(boardId).get();
        if (null == board) return null;
        return board.getCards().values().stream().filter(card -> card.getCardType() == cardType).toList();
    }

    void removeUserFromBoard(String boardId,String email){
        this.findBoardById(boardId).get().getAssignedUsers().remove(email);
    }
}

File: kotlin/complete/myretro/src/main/kotlin/com/apress/myretro/retro/RetroBoardService.kt
package com.apress.myretro.retro

import com.apress.myretro.client.UserClient
import com.apress.myretro.persistence.Repository
import com.apress.myretro.user.User
import org.springframework.stereotype.Service
import java.util.*

@Service
class RetroBoardService(val repository: Repository<RetroBoard, UUID>, val userClient: UserClient) {

    fun getAllAvailableBoards(): Iterable<RetroBoard> = this.repository.findAll()

    fun saveBoard(retroBoard: RetroBoard): RetroBoard = this.repository.save(retroBoard)

    fun findBoardById(boardId:String):RetroBoard? = this.repository.findById(UUID.fromString(boardId))

    fun deleteBoardById(boardId: String) = this.repository.delete(UUID.fromString(boardId))

    fun getCardsFromBoardId(boardId:String): Iterable<Card>? {
        val board = findBoardById(boardId) ?: return emptyList()
        return board.cards?.values
    }

    fun removeCardFromBoard(cardId:String, boardId: String){
        val board = findBoardById(boardId) ?: return
        board.cards?.remove(UUID.fromString(cardId))
    }

    fun saveCardTo(boardId: String, card: Card) : Card? {
        val retroBoard = findBoardById(boardId) ?: return null

        if (null == card.id) {
            val uuid = UUID.randomUUID()
            card.id = uuid
            retroBoard.cards?.set(uuid, card)
            return card
        }

        val updatedCard = retroBoard.cards?.get(card.id)
        updatedCard!!.comment = card.comment
        return updatedCard
    }

    fun assignUserToBoard(boardId: String, email: String): User?{
        val board = findBoardById(boardId) ?: return null
        val user = this.userClient.getUserByEmail(email)

        if (user != null) board.assignedUsers?.put(email,user)

        return user
    }

    fun getHappyCardsFromRetroBoardId(boardId: String?): Iterable<Card?>? {
        return getCardsFromBoardIdAndCardType(boardId!!, CardType.HAPPY)
    }

    fun getMehCardsFromRetroBoardId(boardId: String?): Iterable<Card?>? {
        return getCardsFromBoardIdAndCardType(boardId!!, CardType.MEH)
    }

    fun getSadCardsFromRetroBoardId(boardId: String?): Iterable<Card?>? {
        return getCardsFromBoardIdAndCardType(boardId!!, CardType.SAD)
    }

    fun getCardsFromBoardIdAndCardType(boardId:String, cardType: CardType): Iterable<Card>? {
        val board = findBoardById(boardId) ?: return emptyList()
        return board.cards?.values?.filter { card: Card -> card.cardType == cardType}
    }

    fun removeUserFromBoard(boardId: String?, email: String?) {
        findBoardById(boardId!!)?.assignedUsers?.remove(email)
    }

}